src/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -271,9 +271,7 @@ private static unsafe void CopyImpl(Array sourceArray, int sourceIndex, Array de
             if (destinationArray == null)
                 throw new ArgumentNullException("destinationArray");
 
-            int sourceRank = sourceArray.Rank;
-            int destinationRank = destinationArray.Rank;
-            if (sourceRank != destinationRank)
+            if (sourceArray.EETypePtr != destinationArray.EETypePtr && sourceArray.Rank != destinationArray.Rank)
```

Comment:
`CopyImpl` is a slow path that is not worth optimizing. The case of matching EETypes should be handled earlier by the `RuntimeImports.TryArrayCopy` path - that already contains this optimization.

It make sense for the BlockCopy in your other PR. It should compare `RawValue`.